yktrace/src/hwt/mapper.rs

```diff
@@ -46,54 +50,124 @@ impl BlockMap {
                 crsr.seek(SeekFrom::Current(1)).unwrap();
 
-                map.insert(
-                    usize::try_from(f_off + b_off).unwrap(),
-                    BlockMapEntry { f_off, bb },
-                );
+                let lo = f_off + b_off;
+                let hi = lo + b_sz;
+                elems.push((
+                    (lo..hi),
+                    BlockMapEntry {
+                        f_off,
+                        bb: usize::try_from(bb).unwrap(),
+                    },
+                ));
             }
         }
-        Self { map }
+        Self {
+            tree: elems.into_iter().collect::<IntervalTree<_, _>>(),
+        }
     }
 
     pub fn len(&self) -> usize {
-        self.map.len()
+        self.tree.iter().count()
+    }
+
+    /// Queries the blockmap for blocks whose address range coincides with `start_off..end_off`.
+    fn query(
+        &self,
+        start_off: u64,
+        end_off: u64,
+    ) -> intervaltree::QueryIter<'_, u64, BlockMapEntry> {
+        self.tree.query(start_off..end_off)
     }
 }
 
 pub struct HWTMapper {
-    phdr_offset: u64,
+    symb: Symbolizer,
 }
 
 impl HWTMapper {
     pub(super) fn new() -> HWTMapper {
-        let phdr_offset = get_phdr_offset();
-        HWTMapper { phdr_offset }
+        Self {
+            symb: Symbolizer::new(),
+        }
     }
 
     /// Maps each entry of a hardware trace back the IR block from whence it was compiled.
     pub(super) fn map_trace(&self, trace: Box<dyn Trace>) -> Result<Vec<IRBlock>, HWTracerError> {
-        let blocks = Vec::new();
-        for block in trace.iter_blocks() {
+        let mut ret_irblocks = Vec::new();
+        let mut itr = trace.iter_blocks();
+        while let Some(block) = itr.next() {
             let block = block?;
-
-            // FIXME If the mapping using LLVM is precise, we might be able to get rid of end_addr?
-            let _start_addr = usize::try_from(block.first_instr() - self.phdr_offset).unwrap();
-            let _end_addr = usize::try_from(block.last_instr() - self.phdr_offset).unwrap();
-            todo!(); // FIXME use LLVM mapping info to find the right block in IR.
+            let irblocks = self.map_block(&block);
+            if !ret_irblocks.is_empty() && irblocks.is_empty() {
+                // Once we have seen the last block that can be mapped we are done.
+                break;
+            } else {
+                ret_irblocks.extend(irblocks);
+            }
         }
-        Ok(blocks)
+        // No remaining blocks in the iterator should be mappable. If any can be, then we have
+        // unmappable blocks in the middle of the trace and something is wrong.
+        debug_assert!(itr.all(|block| self.map_block(&block.unwrap()).is_empty()));
+
+        Ok(ret_irblocks)
     }
-}
 
-/// Extract the program header offset.
-fn get_phdr_offset() -> u64 {
-    (&objects()[0]).addr() as u64
+    /// Maps one PT block to one or many LLVM IR blocks.
```

Comment:
This isn't quite what I was expecting. I *think* the example below suggests that LLVM optimised away the implicit `JMP` at the end of `bbx` so that PT sees only one basic block?

If so, I think the example might be more clear if we spell this out. Something like:

```
Imagine we have two optimised LLVM blocks:

bb0:
  x
  y
  JMP bb1
bb1:
  z

If those two basic blocks are compiled adjacent to each other the `JMP bb1` is unneeded so we end up with machine code:

bb0:
  x
  y
bb1:
  z

If the machine code happens to JMP from non-bb0 to bb1 then PT will see a block starting at bb1. However, if execution flows straight from bb0 to bb1 PT will only see bb0. We thus have to expand out the implicit jmp.
```

That's not phrased very well, but hopefully something like that might work?

Reply:
Yes, LLVM has lowered a jump to a fall-thru during code-gen. But apparently that's not how the blockmap section sees it.

During development I had an assert that checked that each blockmap query only returned one block. The assert was failing due to the reason outlined in the docstring.

[Note that in LLVM-IR there is no such concept as fall-thru. All blocks must be properly terminated.]

Reply:
OK, so I think we need to clarify the comment, perhaps as I suggested above: it took me quite a while to work out what's going on. [Possibly because I'm an idiot, but idiots like me do deserve sympathy! The arrow in the comment was a bit misleading IMHO.]

Reply:
I eventually understood the distinction between my comment and yours.

I went back and re-checked the exact reason for the block split.

In one of the mapper tests we have this C func:
```
__attribute__((noinline))                                                                                                       
int                                                                                                                             
add_one_or_two(int arg)                                                                                                         
{                                                                                                                               
    if (arg % 2 == 0) {                                                                                                         
        arg++;                                                                                                                  
    } else {                                                                                                                    
        arg += 2;                                                                                                               
    }                                                                                                                           
    return arg;                                                                                                                 
}         
```

And here's the LLVM IR:
```
; Function Attrs: noinline nounwind optnone uwtable                                                                                                                                                                 
define internal fastcc i32 @add_one_or_two(i32 %0) unnamed_addr #0 {                                                                                                                                                
  %2 = alloca i32, align 4                                                                                                                                                                                          
  store i32 %0, i32* %2, align 4                                                                                                                                                                                    
  %3 = load i32, i32* %2, align 4                                                                                                                                                                                   
  %4 = srem i32 %3, 2                                                                                                                                                                                               
  %5 = icmp eq i32 %4, 0                                                                                                                                                                                            
  br i1 %5, label %6, label %9                                                                                                                                                                                      
                                                                                                                                                                                                                    
6:                                                ; preds = %1                                                                                                                                                      
  %7 = load i32, i32* %2, align 4                                                                                                                                                                                   
  %8 = add nsw i32 %7, 1                                                                                                                                                                                            
  store i32 %8, i32* %2, align 4                                                                                                                                                                                    
  br label %12                                                                                                                                                                                                      
                                                                                                                                                                                                                    
9:                                                ; preds = %1                                                                                                                                                      
  %10 = load i32, i32* %2, align 4                                                                                                                                                                                  
  %11 = add nsw i32 %10, 2                                                                                                                                                                                          
  store i32 %11, i32* %2, align 4                                                                                                                                                                                   
  br label %12                                                                                                                                                                                                      
                                                                                                                                                                                                                    
12:                                               ; preds = %9, %6                                                                                                                                                  
  %13 = load i32, i32* %2, align 4                                                                                                                                                                                  
  ret i32 %13                                                                                                                                                                                                       
}
```

Which is compiled to:
```
            ;-- add_one_or_two:                                                                                                 
            0x00201a10      55             push rbp                                                                             
            0x00201a11      4889e5         mov rbp, rsp                                                                         
            0x00201a14      897dfc         mov dword [rbp - 4], edi                                                             
            0x00201a17      8b45fc         mov eax, dword [rbp - 4]                                                             
            0x00201a1a      b902000000     mov ecx, 2                                                                           
            0x00201a1f      99             cdq                                                                                  
            0x00201a20      f7f9           idiv ecx                                                                             
            0x00201a22      83fa00         cmp edx, 0                                                                           
        ┌─< 0x00201a25      750b           jne 0x201a32                                                                         
        │   0x00201a27      8b45fc         mov eax, dword [rbp - 4]                                                             
        │   0x00201a2a      83c001         add eax, 1                                                                           
        │   0x00201a2d      8945fc         mov dword [rbp - 4], eax                                                             
       ┌──< 0x00201a30      eb09           jmp 0x201a3b                                                                         
       │└─> 0x00201a32      8b45fc         mov eax, dword [rbp - 4]    ; first hit                                 
       │    0x00201a35      83c002         add eax, 2                                                                           
       │    0x00201a38      8945fc         mov dword [rbp - 4], eax                                                             
       └──> 0x00201a3b      8b45fc         mov eax, dword [rbp - 4]    ; second hit                                  
            0x00201a3e      5d             pop rbp                                                                              
            0x00201a3f      c3             ret                                                                                  
```

And:
 * PT saw a block `201a32..=201a3f`.
 * The blockmaper returned 2 hits (which I've annotated in the asm above):
    * `201a32`
    * `201a3b`

So it's actually a mix of both of our reasoning:

 * Yes, there's a fallthru at `201a38` into `201a3b`.
 * The fallthru came from an LLVM `br` (unconditional jump)
 * But the fallthru only exists because there's another incoming edge from `201a30` to `201a3b` and this splits the block.

So both of our comments are correct.

Does that make any sense? I'll try to clarify the comment a bit tomorrow.

Reply:
> But the fallthru only exists because there's another incoming edge from 201a30 to 201a3b and this splits the block.

This doesn't seem like a necessary part of the logic to me. But maybe I'm missing something?

Reply:
Put differently, if there hadn't been that incoming edge, then those two blocks could have been one and talk of fallthru is then moot.

You are still correct though because, although pointless, LLVM could keep two mergable blocks separate.

I think we are agreeing, but we see the situation in different, but equivalent, ways.

Reply:
> Put differently, if there hadn't been that incoming edge, then those two blocks could have been one block and talk of fallthru is then moot.

I don't think that's true. Because even though we only see 1 PT block, when the code-gen removes a branch to the immediately-following basic block, we still need to realise that it represents two LLVM IR blocks.

Reply:
You can see it that way.

If the initial LLVM IR was:
```
bb1:
  ...
  br bb2
bb2:
  ...
```

And the `br` was removed for fallthru, then yes, we'd map one PT block to two IR blocks.

What I'm saying is, I bet even with `-O0` LLVM would merge those.

Regardless, we should be able to handle the case where the blocks are separate.

Reply:
> What I'm saying is, I bet even with -O0 LLVM would merge those.

I don't think it matters about optimisation levels :) I *think* the only important point is that something subsequent to IR optimisation (i.e. the machine code generation part) can remove branches for adjacent blocks. My claim is that whether that's `-O0` or `-O3` or whether there is an incoming edge is irrelevant.

Reply:
Having sat on this overnight, I think I agree that the incoming edge, whilst related, is a distraction. I'll revise the comment.

Reply:
Try 6e8c6111

Reply:
Works for me.